src/output-json-dns.c

```diff
@@ -246,138 +246,52 @@ typedef struct LogDnsLogThread_ {
-{
-    if (!SCDnsLogAnswerEnabled(txptr, LOG_ALL_RRTYPES)) {
-        return NULL;
-    } else {
-        JsonBuilder *js = jb_new_object();
-        SCDnsLogJsonAnswer(txptr, LOG_ALL_RRTYPES, js);
-        jb_close(js);
-        return js;
-    }
-}
-
 bool AlertJsonDns(void *txptr, JsonBuilder *js)
 {
-    bool r = false;
     jb_open_object(js, "dns");
-    JsonBuilder *qjs = JsonDNSLogQuery(txptr);
-    if (qjs != NULL) {
-        jb_set_object(js, "query", qjs);
-        jb_free(qjs);
-        r = true;
-    }
-    JsonBuilder *ajs = JsonDNSLogAnswer(txptr);
-    if (ajs != NULL) {
-        jb_set_object(js, "answer", ajs);
-        jb_free(ajs);
-        r = true;
-    }
+    bool r = SCDnsLogJson(
+            txptr, LOG_FORMAT_DETAILED | LOG_QUERIES | LOG_ANSWERS | LOG_ALL_RRTYPES, js);
     jb_close(js);
     return r;
 }
 
-static int JsonDnsLoggerToServer(ThreadVars *tv, void *thread_data,
-    const Packet *p, Flow *f, void *alstate, void *txptr, uint64_t tx_id)
+static int JsonDnsLogger(ThreadVars *tv, void *thread_data, const Packet *p, Flow *f, void *alstate,
+        void *txptr, uint64_t tx_id)
 {
     SCEnter();
 
     LogDnsLogThread *td = (LogDnsLogThread *)thread_data;
     LogDnsFileCtx *dnslog_ctx = td->dnslog_ctx;
 
-    if (unlikely(dnslog_ctx->flags & LOG_QUERIES) == 0) {
-        return TM_ECODE_OK;
-    }
-
-    for (uint16_t i = 0; i < 0xffff; i++) {
-        JsonBuilder *jb = CreateEveHeader(p, LOG_DIR_FLOW, "dns", NULL, dnslog_ctx->eve_ctx);
-        if (unlikely(jb == NULL)) {
+    if (SCDnsTxIsRequest(txptr)) {
+        if (unlikely(dnslog_ctx->flags & LOG_QUERIES) == 0) {
             return TM_ECODE_OK;
         }
-
-        jb_open_object(jb, "dns");
-        if (!SCDnsLogJsonQuery(txptr, i, td->dnslog_ctx->flags, jb)) {
-            jb_free(jb);
-            break;
+    } else if (SCDnsTxIsResponse(txptr)) {
+        if (unlikely(dnslog_ctx->flags & LOG_ANSWERS) == 0) {
+            return TM_ECODE_OK;
         }
-        jb_close(jb);
-
-        OutputJsonBuilderBuffer(jb, td->ctx);
-        jb_free(jb);
     }
 
-    SCReturnInt(TM_ECODE_OK);
-}
-
-static int JsonDnsLoggerToClient(ThreadVars *tv, void *thread_data,
-    const Packet *p, Flow *f, void *alstate, void *txptr, uint64_t tx_id)
-{
-    SCEnter();
-
-    LogDnsLogThread *td = (LogDnsLogThread *)thread_data;
-    LogDnsFileCtx *dnslog_ctx = td->dnslog_ctx;
+    if (!SCDnsLogEnabled(txptr, td->dnslog_ctx->flags)) {
+        return TM_ECODE_OK;
+    }
 
-    if (unlikely(dnslog_ctx->flags & LOG_ANSWERS) == 0) {
+    JsonBuilder *jb = CreateEveHeader(p, LOG_DIR_FLOW, "dns", NULL, dnslog_ctx->eve_ctx);
+    if (unlikely(jb == NULL)) {
         return TM_ECODE_OK;
     }
 
-    if (SCDnsLogAnswerEnabled(txptr, td->dnslog_ctx->flags)) {
-        JsonBuilder *jb = CreateEveHeader(p, LOG_DIR_FLOW, "dns", NULL, dnslog_ctx->eve_ctx);
-        if (unlikely(jb == NULL)) {
-            return TM_ECODE_OK;
-        }
+    jb_open_object(jb, "dns");
```

Comment:
Could we move this to `SCDnsLogJson` ?
The other protocols do that, so that it is easier to share code for alerts I think...

Reply:
Yes, and done.